Space Turkeys

DaveKorhal opened the pull request:
Adds mini-antags: space turkeys.

<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## What Does This PR Do
Adds mini-antags: space turkeys.
Adds space turkeys event.
<!-- Include a small to medium description of what your PR changes. Document every changes or this may delay review or even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
An event for the Thanksgiving season. Space turkeys are fast, have a leap attack that can knock people down, can lay eggs that hatch into more space turkeys, and gobble a lot. Imagine a less-lethal, more-annoying version of the xeno hunter.
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Images of changes
<!-- If you did not make a map or sprite edit, you may delete this section. You may include a gif of your feature if you want -->

## Changelog
:cl:
add: Space Turkeys miniantags
add: Turkey Infestation event
soundadd: Gobble
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

DaveKorhal commented:
Going to need balance testing, of course.

Known bug: Space Turkeys' Flying Tackle occasionally won't crash into someone in the way, due to the player moving in mid-air. Need a way to stop the player from manually moving in flight.

Future modification: I want to have space turkeys speak vox-pidgin by default, but for some reason languages aren't playing well with simple_animals.

Carthusia commented:
Oh please tell me their sound is the sound they make from southpark

https://www.youtube.com/watch?v=k6HwivfozIQ

Crazylemon64 commented on code/game/gamemodes/miniantags/space_turkey/space_turkey.dm:
```diff
@@ -0,0 +1,117 @@
+	if(!src || QDELETED(src))
+		return
+	var/be_turkey = alert("Become a space turkey? (Warning, You can no longer be cloned!)",,"Yes","No")
+	if(be_turkey == "No" || !src || QDELETED(src))
```
what would happen if someone:
1. clicked the antaghud button
1. clicked the turkey
1. accepted the antaghud prompt, got the information they sought
1. accepted the turkey prompt, with the antag knowledge from antaghud

DaveKorhal replied:
I don't know. I copied that part from the Borer code. If it's an issue with the space turkeys, it'd probably be an issue with the borers as well.

Crazylemon64 replied:
You should test this during the test merge, then, and fix it if that is indeed the case.

DaveKorhal replied:
Tested and fixed.

Note to self: fix the same bug in the borer code.

Crazylemon64 commented on code/game/gamemodes/miniantags/space_turkey/space_turkey.dm:
```diff
@@ -0,0 +1,117 @@
+			notify_ghosts("A space turkey is ready to hatch in [get_area(src)]!", enter_link = "<a href=?src=[UID()];ghostjoin=1>(Click to enter)</a>", source = src, action = NOTIFY_ATTACK)
+			STOP_PROCESSING(SSobj, src)
+	else
+		STOP_PROCESSING(SSobj, src)
```
So if anyone ever grabbed the egg, the turkey egg would become forever dormant?

Crazylemon64 replied:
(this is not a request for a change, but I want confirmation that this is intentional/expected behavior)

DaveKorhal replied:
It shouldn't. At that point the egg is open to be controlled by any ghost. The moment a ghost accepts controlling the egg, the egg is destroyed and it turns into a player-controlled turkey.

Crazylemon64 replied:
The egg would stop ticking once its loc is no longer a turf, e.g. when someone grabs it. I see no mechanism for it to resume ticking.

DaveKorhal replied:
Ah, I misunderstood you. Then yes, as-is it'd become forever dormant. That might be an acceptable bug to ship with, but I'll look into fixing that.

DaveKorhal replied:
I fixed it so eggs grow constantly now. If an egg hatches in a container/backpack/etc, the turkey appears on the floor outside of the container. This causes a weird edge case where if an egg hatches while traveling through Disposals, the turkey will appear right above the Disposals pipe, but since the timing for that is pretty narrow and it just results in a turkey seemingly appearing out of nowhere, I think it's acceptable to leave that in.

Crazylemon64 commented on code/game/gamemodes/miniantags/space_turkey/space_turkey_abilities.dm:
```diff
@@ -0,0 +1,103 @@
+#define MAX_TURKEY_TACKLE_DIST 6
+
+/obj/effect/proc_holder/spell/targeted/turkey_tackle
```
Hmm, do you think it's better like this, or as a subtype of the existing leap spell in goon_powers.dm, perhaps modifying behaviors via overriding certain variables and handlers?

DaveKorhal replied:
I thought about that, but I couldn't think of a good way to append the whole "crashing into someone" bit. Not to mention the turkey's leap temporarily turns on MOBPASS, something the goon's Leap doesn't do.

Crazylemon64 replied:
You could add a `on_collision` method to the leap, as well as a var that when set toggles MOBPASS

DaveKorhal replied:
Perhaps, but unless there's a good advantage to making it a subtype of leap, I'll probably leave it as-is. I know how this one works, and I'm reluctant to break down, remake, and retest the whole thing.

procdrone commented:
Seems somewhat low RP, but otherwise interesting annoyance to deal with.

Bormac commented:
Had an issue while playing as a turkey, it apparently kicked me out mid round, and when I tried to rejoin it was back on the title screen and said I was already logged in. 
According to the admin I spoke with, the turkeys were apparently letting ghosts take control of already controlled turkeys? Not really sure  how that works but either way the server was very angry at me for logging in twice so  I just decided to stay out of that round.
Was interesting while it lasted though. Fun for turkeys, dunno about being on the receiving end of them.

Dashsora commented:
There's a really big issue with taking control of turkeys, you can take over ANY turkey, even already sentient ones (crashing the person who was controlling it, sending them back to the main menu when they reconnect), and dead ones.

Otherwize, really fun - as a turkey.. Should probably write the fact they're not supposed to kill people in **ULTRA BOLD**, or remove their ability to hurt people (Having a lesser tackle/disarm instead of harm would be very nice).

Marcolinoj commented:
I had an issue while trying to enter a turkey's body, the whole screen would freeze, and the third time I tried, it sent me this 
![image](https://user-images.githubusercontent.com/61087399/99313273-ff85b180-283d-11eb-9d68-a8b39caede07.png)

fionalokhert commented:
> Otherwize, really fun - as a turkey.. Should probably write the fact they're not supposed to kill people in **ULTRA BOLD**, or remove their ability to hurt people (Having a lesser tackle/disarm instead of harm would be very nice).

I'll second the notion there, it's listed, but I saw SO many turkeys that I know were sentient killing people.

DaveKorhal commented:
Thanks for the feedback. Gonna be bug-fixing and tweaking for a few days, mainly figuring out how to plug the ghost-hijacking shenanigans.

DaveKorhal commented:
Marcolinoj wrote:
> I had an issue while trying to enter a turkey's body, the whole screen would freeze, and the third time I tried, it sent me this

Looks like you were causing runtime errors for some reason. You remember anything else that happened that might've caused it to act wonky?

For reference, here's the runtime errors Kyet posted that referenced you:
[2020-11-16T21:42:11] Runtime in login.dm,35: Cannot modify null.images.
   proc name: Login (/mob/Login)
   usr: The space turkey (marcolinoj) (/mob/living/simple_animal/hostile/space_turkey)
   usr.loc: The floor (168,139,1) (/turf/simulated/floor/wood)
   src: the space turkey (/mob/living/simple_animal/hostile/space_turkey)
   src.loc: the floor (168,139,1) (/turf/simulated/floor/wood)
   call stack:
   the space turkey (/mob/living/simple_animal/hostile/space_turkey): Login()
   the space turkey (/mob/living/simple_animal/hostile/space_turkey): Login()
   the space turkey (/mob/living/simple_animal/hostile/space_turkey): Login()

[2020-11-16T21:42:11] Runtime in login.dm,4: Cannot read null.address
   proc name: update Login details (/mob/proc/update_Login_details)
   usr: The space turkey (marcolinoj) (/mob/living/simple_animal/hostile/space_turkey)
   usr.loc: The floor (168,139,1) (/turf/simulated/floor/wood)
   src: the space turkey (/mob/living/simple_animal/hostile/space_turkey)
   src.loc: the floor (168,139,1) (/turf/simulated/floor/wood)
   call stack:
   the space turkey (/mob/living/simple_animal/hostile/space_turkey): update Login details()
   the space turkey (/mob/living/simple_animal/hostile/space_turkey): Login()
   the space turkey (/mob/living/simple_animal/hostile/space_turkey): Login()
   the space turkey (/mob/living/simple_animal/hostile/space_turkey): Login()
   the space turkey (/mob/living/simple_animal/hostile/space_turkey): transfer personality(null)
   the space turkey (/mob/living/simple_animal/hostile/space_turkey): attack ghost(the space turkey (/mob/dead/observer))
   the space turkey (/mob/dead/observer): ClickOn(the space turkey (/mob/living/simple_animal/hostile/space_turkey), "icon-x=18;icon-y=12;left=1;scr...")

DevL-Whats-a-jpg commented:
I think the turkeys that were egg spawned or controlled after they spawned didn't get objectives, I at least know I didn't.

Marcolinoj commented:
> Looks like you were causing runtime errors for some reason. You remember anything else that happened that might've caused it to act wonky?

Sorry for the late response. I dont think I did anything to break it, I joined late as an observer, and my connection was normal for sad argentinian connection.

DaveKorhal commented:
Okay, this commit should squash the ghost possession and egg bugs. Also fixed the spawn notices and objectives and edited the objectives. Finally, I added Turkey Infestation to the Moderate Events list as a 0-weighted event. It won't pop off by itself, but it's easier for the admins to activate it.

KozackWolf commented:
Few things turkeys need. They need night vision, cause they can't see at all in a lightless room. Another thing, vent crawl seems to bug out for air vents. More defined limitations in the goals (Don't tackle random people just run around and steal shit, and only tackle if you are cornered or being attacked, or the nest is being broken into.) Longer cooldown on gobble, Give them minor health regen, like a starving terror spider with no way to increase it.

Kepplerx commented:
Absolute hell to fight, they just run around and are like greytiders on speed and meth, unstunnable, with vent crawling

procdrone commented:
vent crawl might be too much.

DaveKorhal commented:
Okay, this is the final build for Thanksgiving. Tweaked egg-laying rate and let eggs constantly grow even if shoved into a container. Turkeys also now take just 40% electrical damage, meaning it takes 3-5 door shocks to kill them instead of just 2-3. (Door shocks were particularly nasty, since the initial shock didn't kill the turkey, which led to them getting shocked twice in 1 second and dropping dead.)

I wanted to add some counterplay to the turkeys' speed involving disablers, bolas, and beartraps, but legcuffs don't work on simple_animals and they have STAMINA damage disabled by default. I'll put them on the To-Do for next Thanksgiving, assuming people like the space turkeys enough to repeat it next year.

(Also spent an hour trying to give them vulp-like darkvision, but the requirements for that seem a bit... wonky.)

Anyway, they're now included as a random Moderate Event with about the same weight as Borer Infestations. Enjoy and Happy Thanksgiving.

Crazylemon64 commented:
conditionally accepted for a test merge over thanksgiving

Necaladun commented:
Absolutely approved for shennigans once off, and love it.

As a regular thing I doubt we'll use it much, it's a bit too silly any other time of the year. But murderturkeys on thanksgiving screams SS13 to me!

A regular turkey mob that isn't any more dangerous than a turkey would also be really nice for less violent thanksgiving events - there was some requests from an admin to be able to do a nicer event with the whole family/dinner/niceness/etc themes of thanksgiving

DaveKorhal commented:
Okay, after seeing another test-merge of the space turkeys in action, here's the changes I plan to make:

**Remove their egg-laying, but spawn 50-100% more of them:** I want the turkeys to feel different from most midround antags. Midround antags like blob, t-spiders, and borers start out small and are either quickly squashed or snowball into giant threats. The turkeys should burst onto the station with the advantages of surprise and numbers, but then slowly be whittled down.

**Reduce their attack damage to 0:** The turkeys have a weak attack, but they can still use it to break machines, computers, and APCs. While it's amusing to watch a bird slowly dismantle the station, there's already plenty of antags that do that. I want to focus on them pestering the crew directly. I might replace it with a pushback or a Disarm so they can still try to defend themselves.

**Let turkeys heal by resting, but doing so reduces their max HP:** I want turkeys to heal up so they don't have to look at red borders the whole time. But I also want them to be slowly whittled down. Resting will heal their HP, but also drop their max HP so they're easier to kill.

variableundefined reviewed (commented):
This was approved for test merge on Thanksgiving, but I object to permanently adding this into the codebase on code quality ground

variableundefined commented on code/game/gamemodes/miniantags/space_turkey/space_turkey.dm:
```diff
@@ -0,0 +1,140 @@
+			attack_ghost(ghost)
+
+/obj/item/reagent_containers/food/snacks/egg/turkey/attack_ghost(mob/user)
+	if(src.amount_grown < TURKEY_HATCH_SPEED)
```
implicit src

variableundefined commented on code/game/gamemodes/miniantags/space_turkey/space_turkey.dm:
```diff
@@ -0,0 +1,140 @@
+	if(href_list["ghostjoin"])
+		var/mob/dead/observer/ghost = usr
+		if(istype(ghost))
+			attack_ghost(ghost)
```
Attack_ghost (Or attack_anything) in general is a function meant to handle something clicking on it. 

Appropriating it for the ghost joining logic is inappropriate.

variableundefined commented on code/game/gamemodes/miniantags/space_turkey/space_turkey.dm:
```diff
@@ -0,0 +1,140 @@
+	if(src.amount_grown < TURKEY_HATCH_SPEED)
+		to_chat(user, "<span class='boldnotice'>This egg is not ready to hatch yet!</span>")
+		return
+	if(!src || QDELETED(src))
```
This check was duplicated three times

variableundefined commented on code/game/gamemodes/miniantags/space_turkey/space_turkey.dm:
```diff
@@ -0,0 +1,140 @@
+	qdel(src)
+
+/mob/living/simple_animal/turkey/space_turkey/attack_ghost(mob/user)
+	if(key)
```
This should be refactored into its own proc to avoid code duplication

variableundefined commented on code/game/gamemodes/miniantags/space_turkey/space_turkey.dm:
```diff
@@ -0,0 +1,140 @@
+	var/be_turkey = alert("Become a space turkey? (Warning, You can no longer be cloned!)",,"Yes","No")
+	if(be_turkey == "No" || !src || QDELETED(src))
+		return
+	if(cannotPossess(user))
```
Why is this check duplicated twice

variableundefined reviewed (commented):
Way too much duplicated code, duplicated checks without explanation for why. Procs belonging to two distinctively different objects are mixed together in a staggered manner.

AffectedArc07 commented:
Object on same reasons as ansari. Was fun for thanksgiving but needs a lot of work to make clean.
